core/src/main/java/org/apache/iceberg/jdbc/JdbcCatalogTransaction.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.jdbc;
+
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.time.Instant;
+import org.apache.iceberg.BaseCatalogTransaction;
+import org.apache.iceberg.BaseTable;
+import org.apache.iceberg.TableMetadata;
+import org.apache.iceberg.Transaction;
+import org.apache.iceberg.TransactionalCatalog;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.exceptions.CommitStateUnknownException;
+import org.apache.iceberg.exceptions.ValidationException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.Tasks;
+
+public class JdbcCatalogTransaction extends BaseCatalogTransaction {
+  private final JdbcClientPool connections;
+
+  public JdbcCatalogTransaction(
+      TransactionalCatalog origin, IsolationLevel isolationLevel, JdbcClientPool connections) {
+    super(origin, isolationLevel);
+    this.connections = connections;
+    beginTransaction();
+  }
+
+  private void beginTransaction() {
+    try {
+      connections.run(
+          conn -> {
+            try (PreparedStatement sql = conn.prepareStatement(JdbcUtil.BEGIN_TX)) {
+              return sql.execute();
+            }
+          });
+    } catch (SQLException | InterruptedException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  @Override
+  public void commitTransaction() {
+    Preconditions.checkState(!hasCommitted(), "Transaction has already committed changes");
+    txByTable().forEach((key, value) -> updatesByTable().put(key, value.pendingUpdates()));
+
+    try {
+      if (IsolationLevel.SERIALIZABLE_WITH_FIXED_READS == isolationLevel()
+          || IsolationLevel.SERIALIZABLE_WITH_LOADING_READS == isolationLevel()) {
+        // make sure tables that are participating in the TX haven't been updated externally
+        for (TableIdentifier affectedTable : updatesByTable().keySet()) {
+          TableMetadata tableMetadata =
+              ((BaseTable) origin().loadTable(affectedTable)).operations().current();
+          if (tableMetadata.lastUpdatedMillis() >= txStartTime().toEpochMilli()) {
```

Comment:
This goes back to my point above, I think checking tableMetadata last updated is too broad when performing the isolation level checks, I think it needs to be performed at the branch level. Unless we're establishing that multi-table transactions can only happen on each table's main branch but that feels too restrictive.

Reply:
same as above, I started working on this before writing to branches was possible, hence I didn't consider this in this initial version